### ISSUE TITLE
Inversion and Conditioning

### ISSUE THREAD

**Issue by zenna** [zenna/OmegaCore.jl#19]
Conditioning is a simple operation (so long as you ignore problems of measure zero).  It means to restrict the sample space to the set of values that are consistent with a predicate.

For example, suppose we have the following normal-normal model:

```julia
μ ~ Normal(0, 1)
x ~ Normal(μ, 1)
μ | x = 5
```
In OmegaCore, you would express this as:

```julia
μ = 1 ~ Normal(0, 1)
x = 2 ~ Normal(μ, 1)
x_ = 0.123
μc = μ |ᶜ x ==ₚ x_
```
There is some amount of syntactic sugar going on here.  The functions constructed in explicit form would be:

```julia
using Distributions: Normal, quantile, cdf

"std-normal inverse cdf"
q(x) = quantile(Normal(0, 1), x)

# Let's assume ω : Int -> [9, 1]
stdunif(id, ω) = ω[id]

n(id, ω, μ_, σ_) = q(stdunif(id, ω)) *  σ_ + + μ_

μ(ω) = n(1, ω, 0, 1)
x(ω) = n(2, ω, μ(ω), 1)
μc(ω) = x(ω) == x_ ? μ(ω) : error()
```

The inference problem is how to generate values according the conditional distribution, which is simply the prior distribution but excluding all those which violate the constraint.

Most inference methods require that we can do two things:
- Generate values consistent with our conditions.  In our case this means generate `ω` such that `x(ω)` is well defined (i.e., does not `error`).
- Know the proposal density of that generated `ω`.

Focusing on the first of these problems first, how can we generate values that satisfy the conditions?
In the example, the output of `x` is conditioned to be `x_ = 5`.  We can then see the following equation:

```
x(ω) = x_ = n(2, ω, μ(ω), 1)
```

Expanding out `n`, we get

```
x_ = q(stdunif(2, ω)) *  1 + μ(ω)
```

Rearranging:

```
(x_ - μ(ω)) / σ_) = q(stdunif(2, ω))
```

And finally, defining `c` as the inverse of `q`, and hence the cdf of the standard normal `c(x) = cdf(Normal(0, 1), x), and substituting in `\sigma = 1` we get:

```
stdunif(2, ω) = c((x_ - μ(ω)) / 1))
```

So, given that `x == x_`, the value of `stdunif(2, ω)`, and hence the value of `ω[2]` is `(x_ - μ(ω)) / 1)`.  This means we have a parametric representation of the subset of the sample space that is consistent with our conditions.

```julia
"return ω such consistent with `x == x_` as parametric function of `v \in [0,1]`"
function consistentω(v, x_)
  # Get value for μ_
  partialω = Dict(1 => v)
  μ_ = μ(partialω)

  # invert
  ω = Dict(1 => v, 2 => c(x_ - μ_))
end
```

Let's plot it:

```julia
using Plots
x_ = 0.1234
ω1(v) = consistentω(v, x_)[1]
ω2(v) = consistentω(v, x_)[2]
plot(ω1, ω2, 0, 1, leg=false)
```

<img width="603" alt="image" src="https://user-images.githubusercontent.com/313946/88217211-34854d00-cc2c-11ea-94d3-46fc5a38ae38.png">

And for `x_ = 2.0`

<img width="589" alt="image" src="https://user-images.githubusercontent.com/313946/88217304-5d0d4700-cc2c-11ea-9244-7f878200aff5.png">

### Proposal Density
As mentioned, being able to generate consistent ω is only half the challenge.  We need to generate consistent value according to the correct distribution.  Many algorithms allow is to do this if we impose a known proposal distribution on the set of consistent ω.

Suppose we place a uniform distribution over `v = ω[1]`.  We might think that since there is a bijection between `v` and `consistentω(v, x_)` (assuming `x_` is fixed) then this would imply a uniform distribution over the set of `consistentω`.  While this would be true if everything were discrete, in the continuous case we have to account for the change of variables.

We could calculate this manually, but also just do it automatically using automatic differentiation through the inverse transform.  Funny thing is, in the case of these primitives, this is equivalent to taking the pdf of the some variable taking some output given its parameters.

### Parametric Inversion
The more general view is this. Suppose we have some random variable `x: Ω -> τ` conditioned on `y: Ω -> Bool`.  We can actually ignore `x`, and see what we really want is:
- To produce some `yinv: Bool × Θ -> Ω`
- Propose `ω` constructively using `yinv(true, θ)`
- Impose distributions on the parameters θ
- Compute the propsal densities on those  proposed ω

The method described above is a certain kind of manually defined parametric inversion for the particular class of primitives.  Actually the parametric inversion infrastructure could probably automate that analysis now, since they don't introduce any of the thorny problems.
In general, parametric inversion is hard though, so one question is then how far can we extend this idea of manual parametric inversions in lieu of automated parametric inversion infrastructure.

One thing we can do is support manual specification of parametric inverses of random variables from their outputs to their parents.
Suppose we have the following model

```julia
x = 1 ~ Normal(0, 1)
y = 2 ~ Normal(0, 1)
z(ω) = x(ω) + y(ω)
z = x + y
xy(ω) = (x(ω), y(ω))
xy |ᶜ z ==ₚ z_
```

Given we know `z_` we can construct a parametric representation of `x` and `y`.

```julia
pinvert(::typeof(z), z_, \theta) =
  (x => z - \theta, y => \theta)
```

It would be quite simple to do this within current infrastructure.

What would we want:
- Avoid creating different and hence inconsistent values of `x` and `y`
- Compute the proposal density on `ω`.  I think Alex's insight is that this can be one compositionally, i.e., without having to define parametric inverse all the way back to primitives.  That seems reasonable, desirable and probably true to me, but I'd need to think about it more

**Comment by zenna** [comment]
Another point is how you specify the local proposals.  In the formulation above you construct the parametric inverse then define parameterisation on theta.  In Alex's formulation you express the proposal distribution on the parents (on `sample` statements in metaprob) directly.  I can see pros and cons of both

**Comment by zenna** [comment]
What would worry me:

Suppose `x` and `y` share `a` as a parent, and both `x` and `y` have defined proposals for `a`.  What happens? Who wins?

**Comment by zenna** [comment]
What would worry me:

suppose `x` and `y` are actually the same but I used the parametric inverse above.  Is that just wrong?  Seems like it

**Comment by alex-lew** [comment]
@zenna Yes. In MetaPPL, you can't run into these problems, because the underlying formulation is different -- generative functions don't get access to some underlying "omega" object.

For example, if `x` and `y` are the same, we would write:

```
def z():
  x = sample(x_dist(), "x")
  return x + x
with proposal(z_):
  return {"x": z_ / 2}
```

The contract for the proposal is to propose a value for each `sample` statement in the model, so there is no way to try and propose different values for the two summands.

Furthermore, if `x` and `y` arise from two different `sample` statements in `z`, they cannot depend on the "same" `a`.

This may be a more fundamental difference between MetaPPL and Omega that lead to (fruitfully!) different approaches to solving this problem.

**Comment by alex-lew** [comment]
> Another point is how you specify the local proposals. In the formulation above you construct the parametric inverse then define parameterisation on theta. In Alex's formulation you express the proposal distribution on the parents (on sample statements in metaprob) directly. I can see pros and cons of both

Right. Another view is that `sample` statements in the `proposal` in MetaPPL _introduce_ dimensions of \theta, and simultaneously place a distribution on them. That is, `sample` needn't directly produce a value that you intend to use as a proposal for a parent; it produces a 'parameter' that you are free to use in producing a proposal for the parent.
